src/dotty/tools/dotc/core/TyperState.scala

```diff
@@ -59,6 +59,19 @@ class TyperState(r: Reporter) extends DotClass with Showable {
   /** Commit state so that it gets propagated to enclosing context */
   def commit()(implicit ctx: Context): Unit = unsupported("commit")
 
+  /** The typer state has already been committed */
+  def isCommitted: Boolean = false
+
+  /** Optionally, if this is a mutable typerstate, it's creator state */
+  def parent: Option[TyperState] = None
+
+  /** The closest ancestor of this typer state (including possibly this typer state itself)
+   *  which is not yet committed, or which does not have a parent.
+   */
+  def uncommittedAncestor: TyperState =
```

Comment:
proposal: instead of storing the direct parent, lazily-cache `uncommittedAncestor` and test-update it every time it's updated:

```
private var lastUncommittedAncestor = parent

def uncommittedAncestor: TyperState = {
  if (lastUncommittedAncestor.isCommited && lastUncommittedAncestor.isDefined) { 
    lastUncommittedAncestor = lastUncommittedAncestor.lastUncommittedAncestor
  }
  lastUncommittedAncestor
}